doc/design/scope.md

```diff
@@ -0,0 +1,58 @@
+# Scope
+
+### Define
+
+Scope is a context to manage Variables. It mainly contains a map from Variable name to Variable. Net will get and update variable throw scope.
+
+```cpp
+class Variable;
+using VariablePtr = std::shared_ptr<Variable>;
+
+class Scope final {
+public:
+  Scope();
+  Scope(const std::shared_ptr<Scope>& parent);
+
+  //! Get Variable in this scope.
+  //! @return nullptr if no such variable.
+  const VariablePtr& getVar(const std::string& name) const;
+
+  //! Create or get a variable in this scope.
+  VariablePtr& createOrGetVar(const std::string& name);
+
+private:
+  /// variable name -> variable
+  std::unordered_map<std::string, VariablePtr> vars_;
+  std::shared_ptr<Scope> parent_{nullptr};
+};
+```
+
+You need to specify a scope to run a Net. One net can run in different scopes and update different variable in the scope. If you did not specify one, It will run in a default scope.
+
+```cpp
+Scope global;
+auto x = newVar("X");  // x is created in scope global, implicitly.
```

Comment:
Just a remind -- the following functions/methods do not following Google naming style:

```
newVar ==> NewVar
addOp ==> AddOp
run ==> Run
```